### ISSUE TITLE
Environment variable for Cargo Workspace

### ISSUE THREAD

**Issue by Ygg01** [rust-lang/cargo#3946]
Hi, while working on using workspace in html5ever, I've ran into issue of needing the `CARGO_WORKSPACE` directory, and being unable, to find it. What I resorted to is essentially, `&Path(cargo_manifest).join("..")` which feels hacky. 

Could `CARGO_WORKSPACE` be added as environment variable? I'm not sure what it should be when there is no workspace defined, I assume it should either return `Err` or default it to `CARGO_MANIFEST_DIR`.

Sidenote I'm willing to work on this issue, if I could get quick pointers, to what I need to do.

**Comment by shepmaster** [comment]
> when there is no workspace defined

I'd expect the environment variable to not be set at all in that case.

**Comment by Ygg01** [comment]
What would adding the CARGO_WORKSPACE entail. From cursory look, I see [custom_build.rs](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/custom_build.rs#L103) has reference to Context, that has reference to Workspace, but same doesn't exist for [compilation.rs](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/compilation.rs#L153).

Would having CARGO_WORKSPACE only for custom builds be ok?

**Comment by alexcrichton** [comment]
Thanks for the report! I think I may not quite be following what's going on here though? Do you mean accessing the workspace directory from a build script perhaps?

**Comment by Ygg01** [comment]
@alexcrichton Yes. I was looking for workspace directory in my custom build script. It's related to servo/html5ever#261. 

 There is a simple workaround of taking `CARGO_MANIFEST_DIR` and using its parent, but I thought having `CARGO_WORKSPACE` would be cleaner solution.

**Comment by alexcrichton** [comment]
Oh yeah definitely makes sense to me! Seems reasonable to basically enhance [this section](http://doc.crates.io/environment-variables.html#environment-variables-cargo-sets-for-build-scripts)

**Comment by Ygg01** [comment]
So if I understand correctly, if I expose [workspace.root_manifest](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/workspace.rs#L33) that would be workspace dir of all projects in workspace, correct? Then I can just:

```rust
if let Some(workspace_dir) = cx.ws.manifest_dir() { 
    cmd.env("CARGO_WORKSPACE_DIR", workspace_dir);
}
```

**Comment by alexcrichton** [comment]
sounds about right! I think you may not want precisely the `root_manifest` field but rather `ws.root().join("Cargo.toml")`

**Comment by Ygg01** [comment]
@alexcrichton I am total newb, but won't adding `Cargo.toml` to workspace root, return the location of workspace manifest as opposed to directory of the directory in which workspace manifest is?

**Comment by alexcrichton** [comment]
oh sure yeah, it just depends on the intent of what's being conveyed (the workspace manifest or the directory of the workspace), I'm fine with either.

**Comment by Ygg01** [comment]
Hm, while writing tests, I've noticed a peculiarity. I assume I'm using this wrong. But I wanted to double check

```rust
    let p = project("foo")
        .file("Cargo.toml", r#"
            [project]
            name = "foo"
            version = "0.5.0"
            authors = []


            [workspace]
            members = ["a"]
        "#)
        .file("src/lib.rs", "")
        .file("build.rs", r#"
            use std::env;
            fn main() {
                //panic!("WILL FAIL");
            }
        "#)
        .file("a/Cargo.toml", r#"
            [project]
            name = "a"
            version = "0.5.0"
            authors = []
            links = "foo"
            build = "build.rs"
        "#)
        .file("a/src/lib.rs", "")
        .file("a/build.rs", r#"
            use std::env;
            fn main() {
                panic!("PASSES?");
            }
        "#);
```

The panic in `a/build.rs` never happens, and panic in `build.rs` always happens, even if workspace doesn't have a `build= "build.rs"` line.

**Comment by alexcrichton** [comment]
@Ygg01 oh `cargo build` on a workspace doesn't execute all build scripts, and `build.rs` is inferred to be a build script if otherwise not specified.

**Comment by Ygg01** [comment]
@alexcrichton Is there an alternative way to test env. variables are properly set in each member build script?

**Comment by alexcrichton** [comment]
@Ygg01 I think you'd just `cargo build` in both directories, right? And then have an assert in the build script the env var is correct?

**Comment by Ygg01** [comment]
Yes, I think that is correct (one call to `cargo build` in member directory and one call in workspace directory). Are there any examples of such code? I can only see adding files to ProjectBuilder file.

**Comment by alexcrichton** [comment]
Oh you'll just want to call `p.cargo` multiple times basically, there's some other tests I believe which run cargo more than once.

**Comment by withoutboats** [comment]
This was assigned to me to summarize why we didn't merge my PR which would have closed it #4787.

We decided to punt on this feature because of the question about what happens when building a crate downloaded from crates.io, which is no longer in a workspace in that form, but might have been originally produced in a workspace and have a build script that expects to have this env var set.

Its also unclear what the motivation for this variable is; my motivation was to find the lockfile, but I concluded that the best way to get the information I was getting from the lockfile was to run `cargo metadata` rather than actually read from Cargo.lock.

**Comment by xftroxgpx** [comment]
> run `cargo metadata` rather than 

hey that's pretty cool:
```bash
$ pwd
/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/recompile_self
$ time cargo metadata --format-version 1 | json_reformat | grep workspace_root
    "workspace_root": "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage"

real	0m1.054s
user	0m0.847s
sys	0m0.214s
```

**Comment by Ygg01** [comment]
@withoutboats my original motivation for this feature was when html5ever, was moving from one project per workspace to multiple. Namely some tests that were specific, became shared and not in the same directory they were left.

However fact that almost no one needed this feature, and it was easily implementable by other means, kinda made me think it's not needed.

**Comment by peterhuene** [comment]
I also have another use case for this feature: I'm trying to get the absolute path to the source file being compiled.  I embed this as metadata so that source can be copied during a subsequent `cargo run` into a particular directory structure to integrate it into a third party, language-agnostic interface.

When using a workspace, the `file!()` macro expands to be workspace directory relative (I was expecting it to be `CARGO_MANIFEST_DIR` relative, though, to be honest).  Without a way to get the workspace directory, I have to rely on a hacky method of walking up the `file!()` path until I hit root, popping sub-directories off of `CARGO_MANIFEST_DIR` as I go.

Of course, there could easily be a better way to get the source file's absolute path that I'm completely ignorant of, so please let me know if that's the case.  Thanks!

**Comment by mitsuhiko** [comment]
I also ran into this problem where `file!` returns a workspace relative path but I cannot absolutize it which I need for some test related situations.

**Comment by mitsuhiko** [comment]
This is the workaround i have in place now which is pretty ugly: https://github.com/mitsuhiko/insta/blob/b113499249584cb650150d2d01ed96ee66db6b30/src/runtime.rs#L67-L88

**Comment by folex** [comment]
Seems like calling `cargo metadata` is the only way currently. Basically what @mitsuhiko done above is
1. Call `cargo metadata --format-version=1`
2. Parsing it as JSON
3. Retrieving field `workspace_root`

So, in a ~nut~shell that would look like
```sh
cargo metadata --format-version=1 | jq .workspace_root
```

I'm a bit concerned about output size of `cargo metadata` though. It's ~1.6MB on my project:
```sh
$ cargo metadata --format-version=1 | wc -c
 1717575
```

**Comment by mitsuhiko** [comment]
This lack of envvar also came back to my mine when investigating this regression: https://github.com/rust-lang/cargo/issues/8992

**Comment by Ygg01** [comment]
So should we just make `CARGO_WORKSPACE_DIR` be convenience for `workspace_root` from `cargo metadata --format-version=1`?

**Comment by hyd-dev** [comment]
> Seems like calling `cargo metadata` is the only way currently. Basically what @mitsuhiko done above is
> 
> 1. Call `cargo metadata --format-version=1`
> 2. Parsing it as JSON
> 3. Retrieving field `workspace_root`
> 
> So, in a ~nut~shell that would look like
> 
> ```shell
> cargo metadata --format-version=1 | jq .workspace_root
> ```
> 
> I'm a bit concerned about output size of `cargo metadata` though. It's ~1.6MB on my project:
> 
> ```shell
> $ cargo metadata --format-version=1 | wc -c
>  1717575
> ```

I think this also works:
```bash
cargo locate-project --workspace --message-format plain
```
The `--workspace` flag is from #8712.

**Comment by pksunkara** [comment]
I think people understand that they can't rely on `CARGO_WORKSPACE_DIR` in `build.rs` because of the reason specified above. But there are lot of other scenarios other than `build.rs` where people want this env var. Would be nice if one of the PRs get accepted and in documentation, we note that using it in `build.rs` would not work.

**Comment by wycats** [comment]
Is there any reason that an environment variable for workspace dir never happened? It seems pretty aligned with the rest of the environment variable and is already exposed via the CLI.

**Comment by RandomInsano** [comment]
Here's a bit of a silly workaround. Create a '[.cargo/config.toml](https://doc.rust-lang.org/cargo/reference/config.html)' file at the root of your workspace and set this as an environment variable:

```
[env]
CARGO_WORKSPACE_DIR = { value = "", relative = true }
```

I don't think this is a good excuse to not have this built in, but at least it's a stop-gap for those who need it. This only works because of the "relative = true" which prepends the path that the ".config" folder is in to the beginning of the variable.

**Comment by Veykril** [comment]
Another use case for this is the `expect-test` crate which has problems in some set ups https://github.com/rust-analyzer/expect-test/issues/33

**Comment by jjant** [comment]
Any updates on this?

**Comment by epage** [comment]
We discussed this today in the cargo team meeting.

tl;dr our starting proposal for this is: A new variable, possibly named `CARGO_MANIFEST_DIR`, would be set for test/bench targets which would help people open paths based on `file!`, likely starting off as relative but the exact format is undefiend atm.  This can start off as a nightly-only PR for us to get a better idea of what we think and then we can evaluate and plan next steps accordingly.

There was the question of whether this is for the "workspace" of what is being built now or the final artifact.  We felt that it would best be associated with what is being built right now, either exposing macros for dependents to use or have them pass paths down into what is being compiled.  That has clearer semantics and doesn't run into inverted dependencies where the crate knows what it is being compiled fir.  This means that for the registry crates, the path would end up being the same as `CARGO_MANIFEST_DIR`.

Someone proposed a theory that making this variable relative would help convey its changeable nature on publish, so we might explore this being a relative path instead of an absolute path.

We were also concerned about making failure cases easier by making this variable available to production code for people accessing files outside of `CARGO_MANIFEST_DIR` which will inaccessible on publish.  We felt that we should start this off as a test/bench only variable, like `CARGO_TARGET_TMPDIR`, but that we can consider expanding the scope in the future based on feedback and use cases.  

One concern raised is that this would primarily be used to make [`std:file!`](https://doc.rust-lang.org/std/macro.file.html) accessible from `current_dir` for opening but `file!` is not fully defined as its intended for debug purposes.  This is another benefit to starting life as test/bench only as it lowers the barrier for abusing `file!`.

The remaining question is on what to name this variable.  Does `CARGO_WORKSPACE_DIR` convey
- Relative path (if we go that route)
- The dynamic meaning of this variable (local vs registry)
- Limited scope of use (at least initially)

**Comment by david-mcgillicuddy-moixa** [comment]
It's funny but we worked around this by using `concat!(env!(CARGO_MANIFEST_DIR), "/..")` in our use case where we invoked `cargo run -p foo` on our own binary from a test, which caused cargo to look up in the parents for a toml, find the workspace toml, and then `-p` works as expected.

Just using `env!(CARGO_MANIFEST_DIR)` as our env failed because it found the _package_ toml and we didn't get proper workspace package resolution for `cargo run -p foo` to work. Having just seen RandomInsano's workaround that's way better.

**Comment by mimoo** [comment]
needed this today again : D

**Comment by epage** [comment]
@yerke in case its useful I hold office hours: https://github.com/rust-lang/cargo/wiki/Office-Hours

**Comment by epage** [comment]
While there might be multiple ways to test this, since a major use case for this is to work with `file!`, we should probably make sure that use case works.

We can test `file!` by having the test define a workspace with a package inside of that workspace.  In that package, we have a test that uses the `file!` macro to assert that a file exists.  Before this feature, that would work by manually specifying how to make the path valid.  After this feature, its just a `Path::join`

**Comment by Nick-Mazuk** [comment]
FYI for those that need this behavior now, I found this crate that gets the current workspace path.

https://crates.io/crates/project-root

Granted an environment variable would be more performance since this crate traverses your file system looking for the nearest Cargo.lock file.

**Comment by weihanglo** [comment]
> Granted an environment variable would be more performance since this crate traverses your file system looking for the nearest Cargo.lock file.

This approach not always correct, as a Cargo.toml can be included/excluded either by itself or from the `[workspace]` section.

There is a pull request getting stale <https://github.com/rust-lang/cargo/pull/12158>. Not sure about if they've got time working on it again, but I guess it pretty acceptable for someone informing them and take over.

**Comment by epage** [comment]
Looking back over use cases.  Unfortunately, there aren't actually that much in the way of details
- `build.rs` wanting to access
  -  servo/html5ever#261
  - https://github.com/rust-lang/cargo/issues/3946#issuecomment-359775523
  - `Cargo.lock` (from #4787)
  - Rejection reason: https://github.com/rust-lang/cargo/issues/3946#issuecomment-359619839
  - Workaround: `cargo metadata`
- Find a binary (https://github.com/rust-lang/cargo/issues/3946#issuecomment-359700772)
  - Seems like this is indirectly seeking #9661
- Resolving `file!()`, particularly for snapshot testing

When the cargo team discussed this (https://github.com/rust-lang/cargo/issues/3946#issuecomment-1402464796) it was mostly focused on `file!()`.

**Comment by epage** [comment]
#12996 provided an alternative to `CARGO_WORKSPACE_DIR` called `CARGO_RUSTC_CURRENT_DIR` which, in theory, should be exactly what `file!` is directly relative to in all cases, rather than indirectly.

rust-lang/rust was updated with this change in rust-lang/rust#118275 (24 Nov 2023) so within a day or two after that, it should be in a nightly.

**Would some of the snapshot testing people be interested in trying this out and providing feedback?**

For other uses, we are continuing to punt, see https://github.com/rust-lang/cargo/issues/3946#issuecomment-359619839

**Comment by lu-zero** [comment]
Another use-case for it is with tools like https://github.com/jamesmunns/toml-cfg since you would like to find the cfg file at the root of the workspace and there is the additional annoyance of having to find the information from a proc_macro.

**Comment by epage** [comment]
@lu-zero so it sounds like that is a case of wanting access to the end-users workspace and not to your own package's workspace.

I think we had expressed concern in the past about being able to change your build with side band information like that.  With https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618, I had considered the idea of allowing something like that though.

**Comment by lu-zero** [comment]
For my specific needs, putting a `cfg.toml` in the root of the workspace is already good. It makes easier to share the same configuration among multiple example-crates that might be radically different (no_std vs std-ish esp targets)

**Comment by epage** [comment]
For the snapshot testing side of this (ie using `std::file!`), I've posted #13644 for stabilizing `CARGO_RUSTC_CURRENT_DIR`.

**Comment by Nemo157** [comment]
Another usecase for this is the [`cargo-xtask` pattern](https://github.com/matklad/cargo-xtask). If you want to have a config file for your xtask then you probably want this stored in the root of the workspace. An xtask crate is not designed to be used outside the workspace so avoids the issues around what happens post-publish.

**Comment by Stargateur** [comment]
My use case is put data for testing in the root of our workspace so I need a simple way to do `include_str!(concat!(env!(CARGO_ROOT_WORKSPACE), "data/foo.json"))` in all my crates and yes I call it `CARGO_ROOT_WORKSPACE` :p

**Comment by epage** [comment]
@Stargateur atm Cargo generally encourages a package to be self-contained, assuming it will be published to crates.io and can be tested from there (like for crater).

I think it would be interesting to evaluate what would be different in Cargo without the published assumption and how we reconcile Cargo functioning under the two workflows.

**Comment by Stargateur** [comment]
I have no strong opinion about what cargo should encourage, I guess focus on "a package should be self contained for published crate" make sense, BUT workspace is not at all about that, for example `tracing.workspace = true` would never be valid for a published crate yet it's in cargo.toml. I think ask for a environnement variable that is the "root" of a project aka workspace is not a big demand. I mean it's asking since 2017... If we have workspace is also because sometime we want share thing between multiple crates.

**Comment by epage** [comment]
>  for example tracing.workspace = true would never be valid for a published crate yet it's in cargo.toml.

That is different as Cargo has enough information to normalize this on publish.

**Comment by bukowa** [comment]
Use case:
```rust
use std::path::Path;

#[track_caller]
fn print_env(){
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    println!("MANIFEST_DIR: {:?}", manifest_dir);

    let caller_file = std::panic::Location::caller().file();
    println!("CALLER FILE: {:?}", caller_file);
    
    let root_dir = Path::new(manifest_dir);
    let this_file = root_dir.join(caller_file);
    println!("RESULT: {:?}", this_file);
}

fn main(){
    print_env()
}
```

**Comment by epage** [comment]
@bukowa Making `file!` usable at runtime is one of the use cases we've been working towards with the unstable `CARGO_RUSTC_CURRENT_DIR` env variable.  There has been debate between Cargo and Libs on which should solve the problem and we need to work to resolve it.

**Comment by epage** [comment]
From #13644

> Discussed this with @Amanieu at RustConf about this.
> 
> In stepping through the use cases, he is now leaning towards `file_absolute!()`. This potentially would warn or error if trim paths is enabled.
> 
> `file!` documentation should be updated to either
> 
>     * unspecified and should only be used for diagnostic output and not for programmatic usage
> 
>     * specified precisely.
> 
> 
> This might need to go through compiler (MCP) and libs-api (ACP)...

As we work towards that, we'll need to remove `CARGO_RUSTC_CURRENT_DIR`.

Another thought I had was that we could change `CARGO_RUSTC_CURRENT_DIR` to always be `CARGO_MANIFEST_DIR` if we fixed the rendering if paths from compiler errors.  However, this would likely make panic messages confusing.

**Comment by epage** [comment]
Created the ACP for `file_absolute!` at rust-lang/libs-team#478

**Comment by teohhanhui** [comment]
No idea how this issue deviated from getting the workspace directory, which for example is very useful for Bevy's asset path, to `file_absolute!` which is entirely useless for that use case...

**Comment by epage** [comment]
@teohhanhui there are multiple use cases discussed in this thread.  A clear cut use case is being able to figure out the runtime path to a source file for snapshot testing and other workflows (https://github.com/rust-lang/cargo/issues/3946#issuecomment-454325570, https://github.com/rust-lang/cargo/issues/3946#issuecomment-1188863621, https://github.com/rust-lang/cargo/issues/3946#issuecomment-412363291).  That is what `file_absolute!` is trying to cover.  I've been focusing on that and posting updates to this thread for a while.

That effort wouldn't close out this issue on its own as there are also other use cases covered here.  For the ones that literally need the workspace root, the situation is less clear cut for moving forward atm:
- One use case is for `toml-cfg` (https://github.com/rust-lang/cargo/issues/3946#issuecomment-1937759335) which is side band communication and its unlikely we'd intentionally do feature work to support this.  See also the next item.
- Accessing workspace-wide data (https://github.com/rust-lang/cargo/issues/3946#issuecomment-2160140148).  This still hasn't been fully explored (why does it need to be workspace-wide? what are the alternatives? etc) and this breaks the assumption that packages are atomic on publish (https://github.com/rust-lang/cargo/issues/3946#issuecomment-359619839).  Granted, not everything gets published but we'd need to approach intentionally supporting features for unpublishable packages carefully
- Find the binary path (https://github.com/rust-lang/cargo/issues/3946#issuecomment-359700772), this is also making assumptions about the target directory location (needing instead #9661) and we do now offer some env variables for looking up binary paths, depending on the use case

**Comment by ratmice** [comment]
In my case I needed to get the path to the workspace for a `.cargo/config.toml` runner.
So that we could set up wasmtime so that the to allow access to all of the paths in the workspace.

It is possible to get at the `workspace_root` key `cargo metadata`, but the environment variable would be much simpler.

**Comment by epage** [comment]
@ratmice would you be able to go into more detail on this?  I'm not quite sure I see how all the pieces fit together.

**Comment by ratmice** [comment]
Yes, I would be happy to,

Say we've got a .cargo/config.toml file like
```
[[target.wasm32-wasip2]
runner = "wasmtime run --dir ."
```

`wasmtime` sandboxes filesystem access such that the current working directory `--dir .` is visible from within wasmtime.

In our case we've got some tests that want to read data files within various crates the repository, these tests are in crates which exist in the workspace, but they aren't crates that get published to crates.io.

When run on `cargo test --target wasm32-wasip2` because we only specified `--dir .` in the `runner`.
These tests can't actually read outside of their working directory.

So I had to write something https://github.com/ratmice/workspace_time_test_runner
which gets the `workspace_root`, and figures out the relative path between the `CARGO_MANIFEST_DIR` 
and the workspace root. Then sets up all of the `--dir` flags between the two (it also keeps wasmtime from sandboxing cargo environment variables).

FWIW, when I wrote my comment some of the most important comments about why the patches for this weren't accepted were hidden. Particularly boats comment here https://github.com/rust-lang/cargo/issues/3946#issuecomment-359619839

**Comment by epage** [comment]
So let's see if I can re-phrase this.

You have target-runner that you want to be able to pass it the path to the entire workspace but using `--dir .` is being evaluated within `CARGO_MANIFEST_DIR` and not a workspace-root, so not enough content is made accessible.  This is so you can access data spread throughout your workspace.

To make sure we have the complete picture, could you provide the motivation for why the files are not consolidated with the tests?  Also, could you use symlinks or are you supporting the Windows without symlinks or wasmtime won't allow resolving of the symlinks?

boats' comment deals with published packages and you had said that yours aren't published.  If it wasn't clear from the other comments you saw, in designing a solution, we have to consider that it "blesses" a workflow and people will use it if they discover it.  We might document there are problems with using it with published packages but documentation tends to be a last resort way of solving a problem as people are unlikely to notice it or even heed it until its too late.  I will say that some day we will need to grapple with the needs of unpublished packages that would be frowned upon within the context of publishing.  I suspect #6153 will be a blocker for that.

**Comment by nuzzles** [comment]
I have a need for this, here's my use-case:

Several binaries.
1 .env file at the root level.

Would love to do this:
```rust
dotenvy::from_path(format!("{}/.env", env!("CARGO_WORKSPACE_DIR"))).ok();
```

**Comment by ratmice** [comment]
@simbleau besides the efficiency of it all/extra process, any reason you cannot just run `cargo metadata`,
something like the following?

```
#[derive(serde::Deserialize)]
struct CargoMetadata {
    workspace_root: std::path::PathBuf,
}

    let output = std::process::Command::new("cargo").args(["metadata"]).output()?;
    let metadata: CargoMetadata = serde_json::from_slice(&output.stdout)?;
    dotenvy::from_path(format!("{}.env", metadata.workspace_root.display()));
```

**Comment by azriel91** [comment]
From a development experience point of view, "why write 7 (fallible) lines when you can write one (infallible)?"

That is, when the spelling of "read the workspace directory" is "run this command, parse the output, format it", it uses more brain power than necessary. 

That said, I haven't read the full thread to see the problems that exist with building individual crates downloaded as dependencies, so certainly room to disregard my comment.

Also, thank you for the snippet, I might turn it into a crate this weekend if someone hasn't already 🙇‍♂️ , that actually might be the right approach for the use cases that *know* when to, and when not to use such a variable.

**Comment by ratmice** [comment]
The main issue with fixing this (from https://github.com/rust-lang/cargo/issues/3946#issuecomment-359619839) 
is that crates.io packages get disembodied from their workspace, leading to potential cases where build.rs works when built from a workspace but not from a package.

Surely it isn't ideal, but it also works today, and doesn't suffer from that specific issue.

**Comment by nuzzles** [comment]
> [@simbleau](https://github.com/simbleau) besides the efficiency of it all/extra process, any reason you cannot just run `cargo metadata`, something like the following?
> 
> ```
> #[derive(serde::Deserialize)]
> struct CargoMetadata {
>     workspace_root: std::path::PathBuf,
> }
> 
>     let output = std::process::Command::new("cargo").args(["metadata"]).output()?;
>     let metadata: CargoMetadata = serde_json::from_slice(&output.stdout)?;
>     dotenvy::from_path(format!("{}.env", metadata.workspace_root.display()));
> ```

My POV is that it's annoying to do this.

If there was a small crate that abstracted this away I'd be happy.

**Comment by ratmice** [comment]
Might as well just use https://crates.io/crates/cargo_metadata if you want a crate.

**Comment by epage** [comment]
We discussed this again in today's cargo team meeting, going over the summary of use cases (https://github.com/rust-lang/cargo/issues/3946#issuecomment-2468442091, https://github.com/rust-lang/cargo/issues/3946#issuecomment-1820152830)
- Accessing `file!` paths at runtime is split out into #15885.
- For a discussion on accessing `target-dir`, that is being discussed at #9661
- For dependencies accessing assets in your workspace: this is not under consideration by the team as we do not want to allow side-banned communication between dependencies and dependents
- For finding the lockfile, a new issue would be needed (note that we have unstable support for overriding what lockfile is used)

The remaining use case that I'm aware of for this issue is "sharing of workspace-wide assets".  I'll be marking this issue as "needs design" within that use case.

Some considerations if anyone wants to pick up this work
- As giving dependencies access to the workspace users are interacting with is a non-starter, the design should avoid allowing this.  If it can't be avoided, a justification should be given as to why
- Workspace-wide assets are inherently incompatible with the notation of extracting a package for publishing to a registry.  Do we limit this feature to only `package.publish = false`?  As its questionable if tests and benches are runnable from a `.crate` (due to workarounds for #4242, lack of verification for all files being included), do we also allow it for tests even if the package can be published?
  - Are these carve outs going to cause enough user confusion / frustration to outweigh the benefits?
  - Is #6153 a blocker?
- What if this variable is being used by a package through a git dependency?  Do we allow it?  If so, I assume that it would need to be the path to that git dependency's own workspace, like us looking it up for workspace inheritance
- What if there is no workspace?  Internally, cargo treats all packages as part of a workspace, its just sometimes implicit

**Comment by Vargas1205** [comment]
VACM880512V89

**Comment by recatek** [comment]
> - As giving dependencies access to the workspace users are interacting with is a non-starter, the design should avoid allowing this. If it can't be avoided, a justification should be given as to why
> - Workspace-wide assets are inherently incompatible with the notation of extracting a package for publishing to a registry.

Say I have a game client and game server that share their assets in a workspace-level `assets/` directory (as Bevy does). I have no intention of ever publishing any crates for my game on crates.io, but I would like to access my workspace root directory to get to that `assets/` directory from it. Is this saying that would never be supported in Cargo?

**Comment by epage** [comment]
No, that was one of the design consideration for if someone picks up designing a solution to this.

**Comment by recatek** [comment]
Gotcha, so if I understand correctly: Putting feasibility aside for a moment, would an approach that populates a `CARGO_WORKSPACE_DIR` environment variable when building crates in the workspace, but not when building dependency crates outside of that workspace satisfy the "giving dependencies access to the workspace users are interacting with is a non-starter" constraint?

**Comment by epage** [comment]
Something along those lines, yes.
